### PR TITLE
Implement LWG-2447: Allocators and volatile-qualified value types

### PR DESCRIPTION
**Pull request by fsb4000**
It doesn't compile already but with unclear error message: https://gcc.godbolt.org/z/G6GE1PaYb

I noticed that when I added a similar patch to LLVM(but about forbid `allocator<const T>`): https://reviews.llvm.org/D125518
(Oh, they intentionally allow `allocator<const T>`, I found this: https://github.com/llvm/llvm-project/commit/a54d028895c91da356a4aaf30e27a5a5b90dd313)

>it caused widespread breakage at Google

:(

gcc is correct and uses `remove_cv` (seen from the error message):
```
error: static assertion failed: std::vector must have a non-const, non-volatile value_type
  435 |       static_assert(is_same<typename remove_cv<_Tp>::type, _Tp>::value,
      |                                                                  ^~~~~
```

https://gcc.godbolt.org/z/3WsT4j3c1

but I think 2 static_asserts with different error messages are better...

### CODE REVIEW

**Comment by fsb4000** [comment]
After reading the LLVM commit I found that LWG-2447 exists

**Comment by frederick-vs-ja** [comment]
Perhaps `polymorphic_allocator` also need these changes.

It seems that `allocator_traits` and `scoped_allocator_adaptor` don't require the template parameters to be allocator types...

**Comment by fsb4000** [comment]
Now, it can break some code.
the example worked [before](https://gcc.godbolt.org/z/r3e4GdxGo):
```c++
#include <array>
#include <list>
#include <memory>
#include <memory_resource>
#include <string>
#include <vector>
using namespace std;
int main() {
    array<std::byte, 32> buffer;
    pmr::monotonic_buffer_resource mbr{ buffer.data(), buffer.size() };
    pmr::polymorphic_allocator<const volatile int> pa{ &mbr };
    pmr::list<const volatile int> list{ pa };
    list.push_front(1);
}
```
now:
```
C:\Dev\STL\out\build\x64\out\inc\xpolymorphic_allocator.h(185): error C2338: static_assert failed: 'The C++ Standard forbids containers of const elements because polymorphic_allocator<const T> is ill-formed.'
main.cpp(12): note: see reference to class template instantiation 'std::pmr::polymorphic_allocator<volatile const int>' being compiled
C:\Dev\STL\out\build\x64\out\inc\xpolymorphic_allocator.h(187): error C2338: static_assert failed: 'The C++ Standard forbids containers of volatile elements because polymorphic_allocator<volatile T> is ill-formed.'
```

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/xmemory
Comment (CaseyCarter) [inline stl/inc/xmemory:788]: FWIW, LWG-2447 doesn't say the default allocator doesn't support cv-qualified types, it says that a type need not support cv-qualified types to be an allocator. I also believe that neither our existing "`allocator<const T>` is ill-formed" claim nor the new "`allocator<volatile T>` is ill-formed" claim is true: https://godbolt.org/z/Ed9e7KhnT.

If we think `allocator` should refuse to support cv-qualified types, I think we should at least file an LWG issue adding wording to the Standard to allow us to do so. We should probably require a non-array object type while we're at it, since the Standard currently places no requirements on `allocator`'s template type parameter.

File: stl/inc/xpolymorphic_allocator.h
Comment (CaseyCarter) [inline stl/inc/xpolymorphic_allocator.h:188]: Ditto "doesn't seem ill-formed to me" https://godbolt.org/z/Tessfzz51

**Comment by fsb4000** [comment]
@CaseyCarter 
I tested gcc behaviour: https://godbolt.org/z/81Me9zbqv

and I found: https://github.com/microsoft/STL/issues/180

So will it be correct if we remove static_asserts from allocators and move them to every container except `map`, `multimap`, `unordered_map`, `unordered_multimap`?

Or should we add them conditionally for C++14 and C++17 only because since C++20 `allocator::address()`  is removed?

https://en.cppreference.com/w/cpp/memory/allocator/address

**Comment by fsb4000** [comment]
I thought a little and did what seemed right to me.

**Comment by frederick-vs-ja** [comment]
I think the changes in this PR should unblock these 2 libcxx tests.

https://github.com/microsoft/STL/blob/b7e0365b5e2f9de738e8fbea42b6cc6880efe531/tests/libcxx/expected_results.txt#L17-L19
https://github.com/microsoft/STL/blob/b7e0365b5e2f9de738e8fbea42b6cc6880efe531/tests/libcxx/skipped_tests.txt#L17-L19

And now I don't think the current title is right:
- LWG-2447 **reduced guarantees**. IIUC (but not sure) it made `Container<volatile T>` UB, and no change is required for implementations.
- This PR **is increasing guarantees**. It is making (possibly) undefined `Container<volatile T>` supported without warning.

Since C++20 `std::allocator<const T>` just fails to meet the _Cpp17Allocator_ requirements, but it needs to be well-formed. So I think this PR should also be labeled with `bug`.

**Comment by strega-nil-ms** [comment]
Given that libc++ intends to remove `allocator<const T>`, and already disallows `allocator<volatile T>`, I don't understand why this PR attempts to add support (message from llvm/llvm-project@a54d028895c91da356a4aaf30e27a5a5b90dd313)

> This reverts commit https://github.com/llvm/llvm-project/commit/276ca87382b8f16a65bddac700202924228982f6. That commit has quite a history at this
point. It was first landed in dbc647643577, which broke std::shared_ptr<T const>
and was reverted in https://github.com/llvm/llvm-project/commit/9138666f5464f53063f1eb44abefbf4c2a0c652e. It was then re-applied in https://github.com/llvm/llvm-project/commit/276ca87382b8f16a65bddac700202924228982f6, with
the std::shared_ptr issue fixed, but it caused widespread breakage at
Google (which suggests it would cause similar breakage in the wild too),
so now I'm reverting again.
>
> Instead, I will add a escape hatch that vendors can turn on to enable
the extension and perform a phased transition over one or two releases
like we sometimes do when things become non-trivial.

**Comment by fsb4000** [comment]
@strega-nil-ms 
Look at https://github.com/microsoft/STL/pull/2719#discussion_r872911363
Yes, I'm fine with both ways. 
We can add `static_assert` to forbid volatile types or if we allow them then we can fix the vector deallocate issue.

**Comment by strega-nil-ms** [comment]
Mmh, I see. Given that libstdc++ allows it, and libc++ does not, I would prefer to take libc++'s approach and disallow it entirely.

**Comment by StephanTLavavej** [comment]
@fsb4000 Would you like to file an LWG issue here, or should we give this to Casey? :joy_cat:

**Comment by fsb4000** [comment]
I don't understand what I should write in that LWG issue, so it would be nice if Casey wrote that issue.

**Comment by StephanTLavavej** [comment]
I just submitted a new LWG issue (no number yet):

---

### `allocator`, `polymorphic_allocator`, and containers should forbid *cv*-qualified types

LWG-2447 adjusted what is now N4944 \[allocator.requirements.general\]/2.1 so that the *Cpp17Allocator* requirements are specified for *cv*-unqualified object types only.

However, nothing in \[default.allocator\] restricts what `T` in `allocator<T>` can be, except where \[allocator.members\]/2 and /6 require it to be complete for `allocate()` and `allocate_at_least()` (and by implication, for `deallocate()`). (Long ago, `allocator` had member functions whose signatures implied that `const`-qualified types were forbidden, but those signatures were deprecated and removed. This explains the phrasing of an MSVC static assertion.)

\[mem.poly.allocator.class.general\] says a bit more about `Tp` in `polymorphic_allocator<Tp>` but doesn't forbid anything. It says that if `Tp` is a *cv*-unqualified object type, then `polymorphic_allocator<Tp>` meets the *Cpp17Allocator* requirements and allocator completeness requirements, but that's all.

There's some implementation variation here. libstdc++ and MSVC reject `allocator<const int>`. libc++ (as of LLVM 16) accepts it but this appears to have been an extension that they're trying to get rid of (their maintainers may be able to comment further; see https://github.com/llvm/llvm-project/commit/a54d028895c91da356a4aaf30e27a5a5b90dd313). These 3 implementations all reject `allocator<volatile int>`, `polymorphic_allocator<const int>`, and `polymorphic_allocator<volatile int>` with varying messages.

The Standard should provide clarity here, by mandating that only *cv*-unqualified object types be given to `allocator`, `polymorphic_allocator`, and containers. (`allocator<void>` must also be allowed, of course. I forget if `polymorphic_allocator` is supposed to accept `void`.) This would simply Standardize existing/desired practice. While it may seem arcane, attempting to form `vector<const T>` is a common novice mistake - so common that MSVC had to add a `static_assert` to emit a clear error message.

---

I was a chicken :baby_chick: and didn't provide a Proposed Resolution.

As this PR was attempting to take the opposite direction, we're going to close it without merging (as @strega-nil-ms explained earlier). Thanks again for looking into this area!

**Comment by frederick-vs-ja** [comment]
IMO it would be more sensible to additionally reject non-object types (except for `void`). Currently `std::allocator<int()>` is supported by the standard and all known implementations since C++20, which makes little sense to me.

> I forget if `polymorphic_allocator` is supposed to accept `void`.

It seems that `polymorphic_allocator<void>` should be supported, despite that it doesn't meet the _Cpp17Allocator_ requirements (see https://github.com/cplusplus/draft/issues/5468#issuecomment-1127284018).

**Comment by StephanTLavavej** [comment]
> IMO it would be more sensible to additionally reject non-object types (except for `void`).

Agreed.

**Pull request closed without merging**
